### PR TITLE
feat: added ability to point to HPCS in different account

### PR DESCRIPTION
**Pull request by Ak-sky**
### Description

Added ability to point to hpcs in different account

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

Added support to attach an IBM Key Management Services (KMS) instance to a VPC cluster from a different account. This requires account_id of the IBM account where KMS instance is deployed. An IBM IAM service to service authorization policy between VPC cluster and KMS instance is required to set up in the IBM account where KMS instance is created.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by Ak-sky** [comment]
Hi @SirSpidey, please review the release notes.

**Comment by SirSpidey** [comment]
@Ak-sky here's a suggestion that fills in a few more details about what the account is different from and what the variable is (or variables are) that signifies the auth policy creation. Let me know what you think.

### Original
> In this release, support is added to point HPCS in different account. For this to work, authorization policy has to be created in the account where the KMS instance is present.

### Suggestion
Added support to point Hyper Protect Crypto Services to a different account than the one specified in xxx. Requires an entry in the `xxx` input variable to create an authorization policy in the account that owns the KMS instance.

**Comment by Ak-sky** [comment]
> @Ak-sky here's a suggestion that fills in a few more details about what the account is different from and what the variable is (or variables are) that signifies the auth policy creation. Let me know what you think.
> 
> ### Original
> > In this release, support is added to point HPCS in different account. For this to work, authorization policy has to be created in the account where the KMS instance is present.
> 
> ### Suggestion
> Added support to point Hyper Protect Crypto Services to a different account than the one specified in xxx. Requires an entry in the `xxx` input variable to create an authorization policy in the account that owns the KMS instance.

Added support to point Hyper Protect Crypto Services to a different account than the one specified in xxx. Requires an entry in the `xxx` input variable to create an authorization policy in the account that owns the KMS instance.

File: variables.tf [hunk 1]
```diff
@@ -168,8 +168,9 @@ variable "kms_config" {
     crk_id           = string
     instance_id      = string
     private_endpoint = optional(bool, true) # defaults to true
+    account_id       = optional(string)     # To attach KMS instance from another account
   })
-  description = "Use to attach a Key Protect instance to the cluster"
+  description = "Use to attach a KMS instance to the cluster"
```
Comment (rajatagarwal-ibm) [inline variables.tf:173]: Description "Use to attach a KMS instance to the cluster. if `account_id` is not provided, defaults to the account in use."

**Comment by rajatagarwal-ibm** [comment]
We should have an example and test (maybe in other_test.go) to demonstrate its working. Need to create a permanent instance in the ops account (subject to confirmation with @vburckhardt).

**Review by rajatagarwal-ibm — CHANGES_REQUESTED** [review]

**Comment by rajatagarwal-ibm** [comment]
Can we have this pointed to the KMS instance rather than just HPCS?

File: variables.tf (hunk 1 above)
Comment (Ak-sky) [reply variables.tf]: Done.

**Comment by Ak-sky** [comment]
> We should have an example and test (maybe in other_test.go) to demonstrate its working. Need to create a permanent instance in the ops account (subject to confirmation with @vburckhardt).

I will work on this once I will get the confirmation, will dicuss this in coming Monday scrum.

**Review by vburckhardt — DISMISSED** [review]

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@Ak-sky I left one comment that you may wish to address in this PR since it has to be rebased with master anyway

**Comment by ocofaigh** [comment]
PS: Remove the term `HPCS` from any release notes here. It can be any supported KMS service (Key Protect or HPCS)

File: modules/fscloud/variables.tf [hunk 2]
```diff
@@ -102,8 +102,9 @@ variable "kms_config" {
     crk_id           = string
     instance_id      = string
     private_endpoint = optional(bool, true) # defaults to true
+    account_id       = optional(string)     # To attach HPCS instance from another account
   })
-  description = "Use to attach a HPCS instance to the cluster"
+  description = "Use to attach a HPCS instance to the cluster. if account_id is not provided, defaults to the account in use."
```
Comment (ocofaigh) [inline modules/fscloud/variables.tf:107]: if -> If

Comment (Ak-sky) [reply modules/fscloud/variables.tf]: Done

**Review by ocofaigh — CHANGES_REQUESTED** [review]
See comments. Also since we are adding a new example, we will want that to onboard to catalog too. Please add the following:

- Add the new example to `ibm_catalog.json`
- Add the new example to `.catalog-onboard-pipeline.yaml`
- Add a `catalogValidationValues.json.template` into the new example directory

File: examples/cross_kms_support/main.tf [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  version = "1.1.4"
+  # if an existing resource group is not set (null) create a new one using prefix
+  resource_group_name          = var.resource_group == null ? "${var.prefix}-resource-group" : null
+  existing_resource_group_name = var.resource_group
+}
+
+########################################################################################################################
+# VPC + Subnet + Public Gateway
+#
+# NOTE: This is a very simple VPC with single subnet in a single zone with a public gateway enabled, that will allow
```
Comment (rajatagarwal-ibm) [inline examples/cross_kms_support/main.tf:16]: I'd put this comment in the Readme of this example.

File: examples/cross_kms_support/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Basic example shows cross account KMS support
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/README.md:1]: Done.

File: examples/cross_kms_support/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+
+A simple example that shows how to provision a basic single zone OCP VPC cluster with cross account KMS encrypttion.
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/README.md:3]: Done.

File: examples/cross_kms_support/README.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+The following resources are provisioned by this example:
+- A new resource group, if an existing one is not passed in.
+- A basic VPC and subnet with public gateway enabled.
+- A single zone OCP VPC cluster.
+- Cross account KMS encryption for OCP VPC cluster.
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/README.md:9]: Done.

File: modules/fscloud/version.tf [hunk 7]
```diff
@@ -3,4 +3,16 @@
...
+    # The below tflint-ignores are required because although the below providers are not directly required by this module,
+    # they are required by consuming modules, and if not set here, the top level module calling this module will not be
+    # able to set alternative alias for the providers.
+    # See https://github.ibm.com/GoldenEye/issues/issues/2390 for full details
+
+    # tflint-ignore: terraform_unused_required_providers
+    ibm = {
+      source  = "ibm-cloud/ibm"
+      version = ">= 1.60.0"
+    }
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply modules/fscloud/version.tf:16]: Done.

**Comment by SirSpidey** [comment]
> @SirSpidey, please review below
> Added support to point KMS instance to a different account. Requires an entry in the kms_config input variable to point the KMS instance to a different account. An authorization policy needs to be created via UI in the account that owns the KMS instance.

**Suggestion**
Added support to point the KMS instance to a different account. Requires an entry in the `kms_config` input variable to point the KMS instance to a different account. Also requires an authorization policy in the account that owns the KMS instance.

- different account: can we say what it's different from? and had suggestions `different account than the one specified in xxx.` 
- auth policy: not clear why `created via UI` is needed in the release notes

File: examples/cross_kms_support/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+output "kms_cross_account_id" {
+  value       = var.kms_cross_account_id
+  description = "Id of the cross account which owns the KMS instance."
+}
+
+output "kms_instance_guid" {
+  value       = var.kms_instance_guid
+  description = "GUID of the KMS instance existing in the cross account."
+}
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/outputs.tf:18]: Done.

File: examples/cross_kms_support/main.tf [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      }
+    ]
+  }
+
+  worker_pools = [
+    {
+      subnet_prefix    = "default"
+      pool_name        = "default" # ibm_container_vpc_cluster automatically names default pool "default" (See https://github.com/IBM-Cloud/terraform-provider-ibm/issues/2849)
+      machine_type     = "bx2.4x16"
+      workers_per_zone = 2 # minimum of 2 is allowed when using single zone
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf:67]: Done.

File: examples/cross_kms_support/main.tf (hunk 3 above)
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf:16]: Done.

File: examples/cross_kms_support/main.tf [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+# NOTE: This is a very simple VPC with single subnet in a single zone with a public gateway enabled, that will allow
+# all traffic ingress/egress by default.
+# For production use cases this would need to be enhanced by adding more subnets and zones for resiliency, and
+# ACLs/Security Groups for network security.
+########################################################################################################################
+
+resource "ibm_is_vpc" "vpc" {
+  name                      = "${var.prefix}-vpc"
+  resource_group            = module.resource_group.resource_group_id
+  address_prefix_management = "auto"
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf:25]: Removed.

**Comment by Ak-sky** [comment]
Added below additional variables in ` catalogValidationValues.json.template` of `cross_kms_support` example.
  "kms_instance_guid": $KMS_INSTANCE_GUID
  "kms_key_crn": $KMS_KEY_CRN
  "kms_cross_account_id": $KMS_ACCOUNT_ID

**Comment by Ak-sky** [comment]
Once Approved, will trigger the PR.

File: examples/cross_kms_support/README.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+- A new resource group, if an existing one is not passed in.
+- A basic VPC and subnet with public gateway enabled.
+- A single zone OCP VPC cluster configured with KMS encryption for cluster data and worker block storage using a KMS encryption key from another account.
+
+
+## NOTE:
+### This is a very simple VPC with single subnet in a single zone with a public gateway enabled, that will allow
+- All traffic ingress/egress by default.
+- For production use cases this would need to be enhanced by adding more subnets and zones for resiliency, and
+- ACLs/Security Groups for network security.
```
Comment (ocofaigh) [inline examples/cross_kms_support/README.md:15]: This seems like it should be part of the previous bullet point?

File: examples/cross_kms_support/catalogValidationValues.json.template [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{
+  "ibmcloud_api_key": $VALIDATION_APIKEY,
+  "region": "au-syd",
+  "resource_tags": $TAGS,
+  "prefix": $PREFIX
+  "kms_instance_guid": $KMS_INSTANCE_GUID
+  "kms_key_crn": $KMS_KEY_CRN
+  "kms_cross_account_id": $KMS_ACCOUNT_ID
```
Comment (ocofaigh) [inline examples/cross_kms_support/catalogValidationValues.json.template:8]: Pipeline is being updated to support `$GE_OPS_ACCOUNT_ID` so please update to use that

File: examples/cross_kms_support/catalogValidationValues.json.template [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+  "ibmcloud_api_key": $VALIDATION_APIKEY,
+  "region": "au-syd",
+  "resource_tags": $TAGS,
+  "prefix": $PREFIX
+  "kms_instance_guid": $KMS_INSTANCE_GUID
```
Comment (ocofaigh) [inline examples/cross_kms_support/catalogValidationValues.json.template:6]: Pipeline is being updated to support `$GE_OPS_KP_US_SOUTH_GUID` so please update to use that

File: examples/cross_kms_support/catalogValidationValues.json.template [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+  "kms_key_crn": $KMS_KEY_CRN
```
Comment (ocofaigh) [inline examples/cross_kms_support/catalogValidationValues.json.template:7]: Pipeline is being updated to support `$GE_OPS_KP_US_ROOT_KEY_CRN` so please update to use that

File: ibm_catalog.json [hunk 15]
```diff
@@ -131,6 +131,23 @@
                             }
                         ]
                     }
+                },
+                {
+                    "label": "Cross account KMS encryption example",
```
Comment (ocofaigh) [inline ibm_catalog.json:136]: remove the word example from here (it shows as example in catalog so no need to add it in title too)

File: ibm_catalog.json [hunk 16]
```diff
@@ -131,6 +131,23 @@
...
+                    "name": "cross_kms_support",
+                    "working_directory": "examples/cross_kms_support",
+                    "architecture": {
+                        "diagrams": [
+                            {
+                                "diagram": {
+                                    "caption": "Cluster security group rules example",
```
Comment (ocofaigh) [inline ibm_catalog.json:143]: caption needs to be updated

**Review by ocofaigh — CHANGES_REQUESTED** [review]
final few comments. Can you also show proof that the test passed locally, since it won't run in this pipeline?

**Comment by Ak-sky** [comment]
> > @SirSpidey, please review below
> > Added support to point KMS instance to a different account. Requires an entry in the kms_config input variable to point the KMS instance to a different account. An authorization policy needs to be created via UI in the account that owns the KMS instance.
> 
> **Suggestion** Added support to point the KMS instance to a different account. Requires an entry in the `kms_config` input variable to point the KMS instance to a different account. Also requires an authorization policy in the account that owns the KMS instance.
> 
> * different account: can we say what it's different from? and had suggestions `different account than the one specified in xxx.`
> * auth policy: not clear why `created via UI` is needed in the release notes 

@ocofaigh @vburckhardt 
 - auth policy: not clear why `created via UI` is needed in the release notes -> This is required because we are not creating service-to-service authorizations between a source service in current IBM Cloud account and a target in the another one via Terraform, so user needs to create from the UI.

Here is another suggestion
- Added support to point KMS instance to a different account other than where the cluster is being deployed. Requires an entry in the kms_config input variable to point the KMS instance to a different account. An authorization policy needs to be created via UI in the account that owns the KMS instance.

File: examples/cross_kms_support/catalogValidationValues.json.template (hunk 12 above)
Comment (Ak-sky) [reply examples/cross_kms_support/catalogValidationValues.json.template:8]: Done.

File: examples/cross_kms_support/catalogValidationValues.json.template (hunk 13 above)
Comment (Ak-sky) [reply examples/cross_kms_support/catalogValidationValues.json.template:6]: Done.

File: examples/cross_kms_support/catalogValidationValues.json.template (hunk 14 above)
Comment (Ak-sky) [reply examples/cross_kms_support/catalogValidationValues.json.template:7]: Done.

File: ibm_catalog.json (hunk 16 above)
Comment (Ak-sky) [reply ibm_catalog.json:143]: Done.

File: ibm_catalog.json (hunk 15 above)
Comment (Ak-sky) [reply ibm_catalog.json:136]: Done.

**Comment by Ak-sky** [comment]
> final few comments. Can you also show proof that the test passed locally, since it won't run in this pipeline?

@ocofaigh Here is the `TestCrossKmsSupportExampleTest` run screenshot-

image

**Comment by SirSpidey** [comment]
> Here is another suggestion
> 
> * Added support to point KMS instance to a different account other than where the cluster is being deployed. Requires an entry in the kms_config input variable to point the KMS instance to a different account. An authorization policy needs to be created via UI in the account that owns the KMS instance.

@Ak-sky I think that works. However, couldn't you create the auth policy with the CLI? (`ibmcloud iam authorization-policy-create`). That's why I'm asking about why the UI needs mentioning.

**Comment by Ak-sky** [comment]
> > Here is another suggestion
> > 
> > * Added support to point KMS instance to a different account other than where the cluster is being deployed. Requires an entry in the kms_config input variable to point the KMS instance to a different account. An authorization policy needs to be created via UI in the account that owns the KMS instance.
> 
> @Ak-sky I think that works. However, couldn't you create the auth policy with the CLI? (`ibmcloud iam authorization-policy-create`). That's why I'm asking about why the UI needs mentioning.

@SirSpidey you are right, will update-

Added support to attach an IBM Key Management Services (KMS) instance to a VPC cluster from a different account. This requires `account_id` of the IBM account where KMS instance is deployed. An IBM IAM service to service authorization policy between VPC cluster and KMS instance is required to set up in the IBM account where KMS instance is created.

File: examples/cross_kms_support/catalogValidationValues.json.template [hunk 17]
```diff
@@ -0,0 +1,9 @@
+{
+  "ibmcloud_api_key": $VALIDATION_APIKEY,
+  "region": "au-syd",
+  "resource_tags": $TAGS,
+  "prefix": $PREFIX
+  "kms_instance_guid": $GE_OPS_KP_US_SOUTH_GUID
+  "kms_key_crn": $GE_OPS_KP_US_ROOT_KEY_CRN
+  "kms_cross_account_id": $GE_OPS_ACCOUNT_ID
```
Comment (ocofaigh) [inline examples/cross_kms_support/catalogValidationValues.json.template:8]: invalid json - missing trailing commas

File: examples/cross_kms_support/main.tf [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+  worker_pools         = local.worker_pools
+  access_tags          = var.access_tags
+
+  kms_config = {
+    instance_id = var.kms_instance_guid
+    crk_id      = local.kp_key_id
+    account_id  = var.kms_cross_account_id
+  }
+
+
```
Comment (ocofaigh) [inline examples/cross_kms_support/main.tf:98]: double blank line

File: examples/cross_kms_support/README.md [hunk 19]
```diff
@@ -0,0 +1,10 @@
+# Cross account KMS encryption example
+
+A simple example that shows how to provision a VPC with single subnet in a single zone with a public gateway enabled OCP VPC cluster with cross account KMS encryption.
+
```
Comment (ocofaigh) [inline examples/cross_kms_support/README.md:4]: we need to add a about having the cross account auth policy set set before running the example

File: examples/cross_kms_support/main.tf (hunk 18 above)
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf:98]: Done.

File: examples/cross_kms_support/main.tf [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+  total_ipv4_address_count = 256
+  public_gateway           = ibm_is_public_gateway.gateway.id
+}
+
+########################################################################################################################
+# OCP VPC cluster (single zone)
+########################################################################################################################
+
+locals {
+  kp_key_id = regex("key:(.*)", var.kms_key_crn)[0]
```
[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf:49]: I was trying to keep it consistent with the other examples- https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/blob/4d91ce91ab96b330f1c20ca67a969485ff8d6b39/examples/fscloud/variables.tf#L46

[the remark this answers is not in the document]
Comment (Ak-sky) [reply examples/cross_kms_support/main.tf]: Changes done.

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**